File: Session_03_DivideAndConquer/multi_language_implementations/divide_and_conquer.cpp
#include <iostream>
#include <vector>

void printArray(int arr[], int size) {
    std::cout << "[";
    for (int i = 0; i < size; i++) {
        std::cout << arr[i];
        if (i < size - 1) std::cout << ", ";
    }
    std::cout << "]\n";
}

int binarySearch(int arr[], int low, int high, int target) {
    if (low > high) return -1;
    int mid = low + (high - low) / 2;
    std::cout << "  Binary Search: Checking index " << mid << " (value: " << arr[mid] << ") in range [" << low << ", " << high << "]\n";
    if (arr[mid] == target) return mid;
    if (arr[mid] > target) return binarySearch(arr, low, mid - 1, target);
    else return binarySearch(arr, mid + 1, high, target);
}

void merge(int arr[], int low, int mid, int high) {
    int leftSize = mid - low + 1;
    int rightSize = high - mid;
    std::vector<int> leftArr(leftSize);
    std::vector<int> rightArr(rightSize);
    for (int i = 0; i < leftSize; i++) leftArr[i] = arr[low + i];
    for (int j = 0; j < rightSize; j++) rightArr[j] = arr[mid + 1 + j];
    int i = 0, j = 0, k = low;
    while (i < leftSize && j < rightSize) {
        if (leftArr[i] <= rightArr[j]) {
            arr[k] = leftArr[i];
            i++;
        } else {
            arr[k] = rightArr[j];
            j++;
        }
        k++;
    }
    while (i < leftSize) {
        arr[k] = leftArr[i];
        i++;
        k++;
    }
    while (j < rightSize) {
        arr[k] = rightArr[j];
        j++;
        k++;
    }
}

void mergeSort(int arr[], int low, int high) {
    if (low < high) {
        int mid = low + (high - low) / 2;
        mergeSort(arr, low, mid);
        mergeSort(arr, mid + 1, high);
        merge(arr, low, mid, high);
        std::cout << "  Merged range [" << low << " to " << high << "]: ";
        for (int x = low; x <= high; x++) std::cout << arr[x] << " ";
        std::cout << "\n";
    }
}

int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;
    return i + 1;
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        std::cout << "  Partition pivot chosen: " << arr[pi] << ". Array state: ";
        printArray(arr, 5);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int main() {
    int searchArr[] = {2, 5, 8, 12, 16, 23, 38, 56, 72, 91};
    int searchSize = sizeof(searchArr) / sizeof(searchArr[0]);
    int target = 23;
    std::cout << "=== PART 1: Recursive Binary Search ===\n";
    int resultIndex = binarySearch(searchArr, 0, searchSize - 1, target);
    if (resultIndex != -1) std::cout << "Target found at index: " << resultIndex << "\n\n";
    else std::cout << "Target not found.\n\n";

    int mergeArr[] = {38, 27, 43, 3, 9, 82, 10};
    int mergeSize = sizeof(mergeArr) / sizeof(mergeArr[0]);
    std::cout << "=== PART 2: Merge Sort ===\n";
    mergeSort(mergeArr, 0, mergeSize - 1);
    std::cout << "\n";

    int quickArr[] = {5, 2, 9, 1, 6};
    int quickSize = sizeof(quickArr) / sizeof(quickArr[0]);
    std::cout << "=== PART 3: Quick Sort ===\n";
    quickSort(quickArr, 0, quickSize - 1);
    std::cout << "Press Enter to exit...";
    std::cin.get();
    return 0;
}

File: Session_03_DivideAndConquer/multi_language_implementations/divide_and_conquer.py
import sys

def print_array(arr, size):
    print("[", end="")
    for i in range(size):
        print(arr[i], end="")
        if i < size - 1:
            print(", ", end="")
    print("]")

def binary_search(arr, low, high, target):
    if low > high:
        return -1
    mid = low + (high - low) // 2
    print(f"  Binary Search: Checking index {mid} (value: {arr[mid]}) in range [{low}, {high}]")
    if arr[mid] == target:
        return mid
    if arr[mid] > target:
        return binary_search(arr, low, mid - 1, target)
    else:
        return binary_search(arr, mid + 1, high, target)

def merge(arr, low, mid, high):
    left_size = mid - low + 1
    right_size = high - mid
    left_arr = arr[low : mid + 1]
    right_arr = arr[mid + 1 : high + 1]
    
    i = 0
    j = 0
    k = low
    while i < left_size and j < right_size:
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1
    
    while i < left_size:
        arr[k] = left_arr[i]
        i += 1
        k += 1
        
    while j < right_size:
        arr[k] = right_arr[j]
        j += 1
        k += 1

def merge_sort(arr, low, high):
    if low < high:
        mid = low + (high - low) // 2
        merge_sort(arr, low, mid)
        merge_sort(arr, mid + 1, high)
        merge(arr, low, mid, high)
        print(f"  Merged range [{low} to {high}]: ", end="")
        for x in range(low, high + 1):
            print(f"{arr[x]} ", end="")
        print()

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        print(f"  Partition pivot chosen: {arr[pi]}. Array state: ", end="")
        print_array(arr, 5)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

def main():
    search_arr = [2, 5, 8, 12, 16, 23, 38, 56, 72, 91]
    search_size = len(search_arr)
    target = 23
    print("=== PART 1: Recursive Binary Search ===")
    result_index = binary_search(search_arr, 0, search_size - 1, target)
    if result_index != -1:
        print(f"Target found at index: {result_index}\n")
    else:
        print("Target not found.\n")

    merge_arr = [38, 27, 43, 3, 9, 82, 10]
    merge_size = len(merge_arr)
    print("=== PART 2: Merge Sort ===")
    merge_sort(merge_arr, 0, merge_size - 1)
    print()

    quick_arr = [5, 2, 9, 1, 6]
    quick_size = len(quick_arr)
    print("=== PART 3: Quick Sort ===")
    quick_sort(quick_arr, 0, quick_size - 1)
    input("Press Enter to exit...")

if __name__ == "__main__":
    main()
